Sources/SwiftyCVPixelBuffer/CVPixelBuffer+Utils.swift

```diff
@@ -75,26 +77,73 @@ public extension CVPixelBuffer {
     }
     
     /// Deep copy a CVPixelBuffer:
-    /// http://stackoverflow.com/questions/38335365/pulling-data-from-a-cmsamplebuffer-in-order-to-create-a-deep-copy
+    /// https://stackoverflow.com/a/58647596/3940420
     func makeCopy() throws -> CVPixelBuffer {
-        let copy = try self.allocateBlankCopy()
-        
-        self.lockBaseAddress(options: .readOnly)
-        copy.lockBaseAddress(options: [])
+        precondition(CFGetTypeID(self) == CVPixelBufferGetTypeID(), "copy() cannot be called on a non-CVPixelBuffer")
+
+        var _copy: CVPixelBuffer?
+
+        let width = CVPixelBufferGetWidth(self)
+        let height = CVPixelBufferGetHeight(self)
+        let formatType = CVPixelBufferGetPixelFormatType(self)
+        let attachments = CVBufferGetAttachments(self, .shouldPropagate)
+
+        CVPixelBufferCreate(nil, width, height, formatType, attachments, &_copy)
+
+        guard let copy = _copy else {
+            throw Error.allocationFailed
+        }
+
+        CVPixelBufferLockBaseAddress(self, .readOnly)
+        CVPixelBufferLockBaseAddress(copy, [])
+
         defer {
-            copy.unlockBaseAddress(options: [])
-            self.unlockBaseAddress(options: .readOnly)
+            CVPixelBufferUnlockBaseAddress(copy, [])
+            CVPixelBufferUnlockBaseAddress(self, .readOnly)
         }
-        
-        // TODO: Handle non-planar pixel buffers
-        for plane in 0 ..< self.planeCount {
-            let dest = copy.baseAddress(of: plane)
-            let source = self.baseAddress(of: plane)
-            let height = self.height(of: plane)
-            let bytesPerRow = self.bytesPerRow(of: plane)
-            memcpy(dest, source, height * bytesPerRow)
+
+        let pixelBufferPlaneCount: Int = CVPixelBufferGetPlaneCount(self)
+
+
+        if pixelBufferPlaneCount == 0 {
+            let dest = CVPixelBufferGetBaseAddress(copy)
+            let source = CVPixelBufferGetBaseAddress(self)
+            let height = CVPixelBufferGetHeight(self)
+            let bytesPerRowSrc = CVPixelBufferGetBytesPerRow(self)
+            let bytesPerRowDest = CVPixelBufferGetBytesPerRow(copy)
+            if bytesPerRowSrc == bytesPerRowDest {
+                memcpy(dest, source, height * bytesPerRowSrc)
+            } else {
+                var startOfRowSrc = source
+                var startOfRowDest = dest
+                for _ in 0..<height {
+                    memcpy(startOfRowDest, startOfRowSrc, min(bytesPerRowSrc, bytesPerRowDest))
+                    startOfRowSrc = startOfRowSrc?.advanced(by: bytesPerRowSrc)
+                    startOfRowDest = startOfRowDest?.advanced(by: bytesPerRowDest)
+                }
+            }
+
+        } else {
+            for plane in 0 ..< pixelBufferPlaneCount {
+                let dest        = CVPixelBufferGetBaseAddressOfPlane(copy, plane)
```

Comment:
strange alignment

Reply:
also, let's use extensions instead of C-style funcs